### DISCUSSION TITLE
Focus not preserved on input when submitting form

### DISCUSSION THREAD

**Discussion by thetutlage** [unpoly/unpoly#658] [General]
I have a very basic form that I am submitting using Unpoly and expected the focus to be preserved on the `input` where I pressed enter to submit the form.

However, it looks like the focus is moved to the `submit` button to trigger the form submission and hence Unpoly re-focuses the same button.

Here's a recording of the same.

https://github.com/user-attachments/assets/c327cadf-ac96-479f-acf4-d5780e9a285c

Do note: The server follows PRG pattern for form submissions and in-case of the error it will trigger a redirect with the errors as flash messages

**Comment by adam12** [comment]
Can you confirm the version of Unpoly you're using? Have you modified the [focus configuration](https://unpoly.com/focus) at all?

**Reply by thetutlage** [reply]
Have not touched any configuration at all. I am using `v3.8.0`

I can create a reproduction tomorrow if that will be helpful

**Reply by triskweline** [reply]
> I can create a reproduction tomorrow if that will be helpful

I can't immediately see why this would happen. A minimal reproduction would be very helpful.

**Reply by thetutlage** [reply]
Sure. Here it is. Pressing the enter button inside the `input` box moves the focus to the `button` element. https://elite-hill-shrimp.glitch.me/

Link to code: https://glitch.com/edit/#!/elite-hill-shrimp

**Reply by triskweline** [reply]
Here's what's happening:

- When the user submits a form by pressing `Enter` in a focused input, [the browser finds the first submit button and fires a `click` event on it](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#implicit-submission). TIL.
- When a submit button is clicked, Unpoly forces focus on that button. This is to remember the submit button so its name/value can go into the request parameters. This fix is only for MacOS, which does by default not focus clicked buttons.

Maybe we can use another way to remember the submit button. The `submit` event seems to have a `{ submitter }` property that might work.

**Reply by triskweline** [reply]
Original MacOS issue: [[Mac] Invalid form params when multiple submit buttons are used](https://github.com/unpoly/unpoly/issues/103)

**Reply by thetutlage** [reply]
Alright, I get it now. Thanks for sharing all these informative links :)

**Comment by triskweline** [comment]
This should be fixed in https://github.com/unpoly/unpoly/commit/e8fcd40a83b64cd09b7c642e98055fe28f2badd4 .

**Reply by triskweline** [reply]
Can't do a full release now, but available as `3.9.0-alpha2` in case this blocks you.

**Reply by thetutlage** [reply]
Chef kiss! Thanks a ton for the quick fix :)

**Comment by triskweline** [comment]
This issue has been resolved with version 3.9.0.
